src/layout.rs

```diff
@@ -32,13 +32,18 @@ pub enum Event {
     Press(u8, u8),
     /// Release event with coordinates (i, j).
     Release(u8, u8),
+    /// Press event with just a keycode (for sequences)
+    SequencePress(KeyCode),
+    /// Release event with just a keycode (for sequences)
+    SequenceRelease(KeyCode),
```

Comment:
You can't modify this enum. It's the public contract of events entering a layout. This enum is not `non_exhaustive`, and thus adding variants is a breaking change.

Reply:
are breaking changes bad for keyboard firmware - I'm build and flash it all in one go so from my side I wouldn't be affected by a breaking change?  - maybe we should break now, up the semver and make it non exhaustive at the same time?

Reply:
Breaking change will say that you don't to just do cargo update to update keyberon. I don't care of making breaking changes, but it must add something to the user.

But here, the problem is more that implementation details leak in the public interface.

Reply:
@gilescope I think TeXitoi was just making a point (in a nice, technical way) that how I implemented this particular feature sucks, haha.  He's right though!  I've got a much better solution that I'm currently working on that should allow us to incorporate specific delays between keystrokes instead of just one universal delay that's the same for all `Press()` and `Release()` events (in a given macro).

I'm still new to Rust and his wisdom is appreciated :smile: .  I'm currently struggling with writing code (a `map()`, `filter()` or `filiter_map()`) that figures out which keycodes need to remain down during the execution of any given sequence.  It's not so simple because if you have a `Press()` but no corresponding `Release()` until the end of the sequence you need to A) figure out where in the sequence you are and B) cancel out matching pairs of `Press()` and `Release()` events *but only if they're within the same window of the sequence* (that you're currently at).

I actually got the code working to execute simple sequences but it currently doesn't work with modifiers or anything that needs to remain held during the sequence.  Hopefully not much longer now!

Aside: In the process of writing this I've learned a lot about how HID keyboards work, haha.  Like how you can only have a maximum of six keys down at a time--even though there's no real technical reason behind this limitation.  It's just that when folks were defining the USB HID spec they didn't say how many keycodes could be sent at once so *vendors* (e.g. Apple and Microsoft) limited it to six simultaneous keycodes in their implementations (or at least, that's how it was explained to me).  No idea what happens if you try to send more but it's in my TODO list of things to play with =)

Reply:
The "usb keyboard boot protocol" has this 6 keys (+8 modifiers) limitation. There is alternative USB descriptor, but the one used in keyberon has this limitation (that I didn't find limiting, so I didn't try the "NKRO" descriptor support).